---
title: "_igblastr_ overview"
author:
- name: Hervé Pagès
  affiliation: Fred Hutch Cancer Center, Seattle, WA
- name: Kellie MacPhee
  affiliation: Fred Hutch Cancer Center, Seattle, WA
date: "Compiled `r BiocStyle::doc_date()`; Modified 11 April 2025"
package: igblastr
vignette: |
  %\VignetteIndexEntry{igblastr overview}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
output:
  BiocStyle::html_document
---


```{r setup, include=FALSE}
library(BiocStyle)
```


# Introduction

The Immunoglobulin Basic Local Alignment Search Tool (IgBLAST) is a specialized bioinformatics tool developed by the National Center for Biotechnology Information (NCBI) for the analysis of B-cell receptor (BCR) and T-cell receptor (TCR) sequences (Ye et al., 2013). IgBLAST performs sequence alignment and annotation, with key outputs including germline V, D, and J gene assignments; detection of somatic hypermutations introduced during affinity maturation; identification and annotation of complementarity-determining regions (CDR1–CDR3) and framework regions (FR1–FR4); and both nucleotide and protein-level alignments. 

`r Biocpkg("igblastr")` is an R/Bioconductor package that provides functions
to conveniently install and run a local IgBLAST installation from within R.
The package is designed to make it as easy as possible to use IgBLAST in R
by streamlining the installation of both IgBLAST and its associated germline 
databases. In particular, these installations can be performed with a single 
function call, do not require root access, and persist across R sessions.

The main function in the package is `igblastn()`, a wrapper to the `igblastn`
_standalone executable_ included in IgBLAST. In addition to `igblastn()`,
the package provides:

- A function (`install_igblast()`) for conveniently downloading and
  installing a pre-compiled IgBLAST from NCBI.

- Functions to download and install germline databases from the
  IMGT/V-QUEST download site, and to configure them for use with `igblastn()`.

- A set of built-in germline databases from OGRDB, the AIRR Community’s Open Germline Reference Database.

- A set of built-in  constant region (C-region) databases (a.k.a. constant region databases) from IMGT/V-QUEST.

- Utility functions to parse the results returned by `igblastn()`.

- Utilities to download data from OAS, the Observed Antibody Space database,
  and prepare it for use with IgBLAST.

- Etc.

IgBLAST is described at https://pubmed.ncbi.nlm.nih.gov/23671333/

Online IgBLAST: https://www.ncbi.nlm.nih.gov/igblast/

Please use https://github.com/HyrienLab/igblastr/issues to report bugs,
provide feedback, request features (etc) about `r Biocpkg("igblastr")`.


# Install and load the package


Like any Bioconductor package, `r Biocpkg("igblastr")` should be installed with `BiocManager::install()`:
```{r, eval=FALSE}
if (!require("BiocManager", quietly=TRUE))
    install.packages("BiocManager")
BiocManager::install("igblastr")
```

`BiocManager::install()` will take care of installing the package dependencies
that are missing.

Load the package:
```{r, message=FALSE}
library(igblastr)
```


# Install IgBLAST


If IgBLAST is already installed on your system, you can tell
`r Biocpkg("igblastr")` to use it by setting environment variable
`IGBLAST_ROOT` to the path of the IgBLAST installation.
See `?IGBLAST_ROOT` for more information.

Otherwise, simply call `install_igblast()` to install the latest
version of IgBLAST. As of March 2025, NCBI provides pre-compiled
versions of IgBLAST for Linux, Windows, Intel Mac and Mac Silicon.
`install_igblast()` will download the pre-compiled IgBLAST that matches
your platform from NCBI FTP site, and install it in a location that will
be remembered across R sessions.
```{r}
if (!has_igblast())
    install_igblast()
```
See `?install_igblast` for more information.

Note that we used `has_igblast()` to avoid trying to reinstall IgBLAST if
`r Biocpkg("igblastr")` has already access to an IgBLAST installation.

Display basic information about the IgBLAST installation used
by `r Biocpkg("igblastr")`:
```{r}
igblast_info()
```


# Install and select a germline database


The `r Biocpkg("igblastr")` package includes a FASTA file that contains
8437 pairs of human antibody sequences (16874 individual sequences)
retrieved from OAS (the Observed Antibody Space database). These will
be our _query sequences_ i.e. the immunoglobulin (IG) sequences that
we're going to analyze in this vignette.

Before we can do so, the `igblastn` _standalone executable_ in IgBLAST (and
thus our `igblastn()` function) will need access to the germline V, D,
and J gene sequences for human. And before `igblastn` can use these
germline sequences, they must be stored in three separate blast databases.
We'll refer to these databases as the V-region, D-region, and J-region
databases. When the three databases are put together, we'll refer to the
aggregated database as the _germline database_.

## Builtin germline databases

The `r Biocpkg("igblastr")` package includes a set of builtin germline
databases for human and mouse that were obtained from the OGRDB database
(AIRR community). These can be listed with:
```{r}
list_germline_dbs(builtin.only=TRUE)
```
The last part of the database name indicates the date of the download in
YYYYMM format.

If our query sequences are from human or from one of the mouse strains listed
above, we can already select the appropriate database with `use_germline_db()`
and skip the subsection below.

## Install additional germline databases

AIRR/OGRDB and IMGT/V-QUEST are two providers of germline databases that
can be used with IgBLAST. If, for whatever reason, using one of the builtin
AIRR/OGRDB germline databases is not an option (e.g. our query sequences
are not from human or mouse), then we can use `install_IMGT_germline_db()`
to install additional germline databases. Here we show how to install the
latest human germline database from IMGT/V-QUEST.

First we list the most recent IMGT/V-QUEST releases:
```{r}
head(list_IMGT_releases())
```

The organisms included in release 202506-1 are:
```{r}
list_IMGT_organisms("202506-1")
```

Then we install the human germline database from the latest IMGT/V-QUEST
release:
```{r, message=FALSE}
install_IMGT_germline_db("202506-1", organism="Homo sapiens", force=TRUE)
```
See `?install_IMGT_germline_db` for more information.

Finally we select this new germline db to use with `igblastn()`:
```{r}
use_germline_db("IMGT-202506-1.Homo_sapiens.IGH+IGK+IGL")
```
See `?use_germline_db` for more information.

## Display the full list of local germline dbs

To see the full list of local germline dbs:
```{r}
list_germline_dbs()
```
Note that the asterisk (`*`) displayed on the very right end
of `list_germline_dbs()`'s output indicates the germline db that
is currently selected (you might need to scroll the output to the
right to see the asterisk).

See `?list_germline_dbs` for more information.


# Select a constant region database (optional)


The `igblastn` _standalone executable_ in IgBLAST can also use C gene
sequences, a.k.a. constant regions, to improve the analysis. Like with the
germline V, D, and J gene sequences, the C sequences should typically be
from the same organism than the query sequences, and they also need to be
stored in a blast database. We'll refer to this database as the C-region db.

The `r Biocpkg("igblastr")` package includes a set of builtin C-region dbs
for human, mouse, and rabbit that were obtained from IMGT/V-QUEST. These
can be listed with:
```{r}
list_c_region_dbs()
```
The last part of the database name indicates the date of the download in
YYYYMM format.

If our query sequences are from human, mouse, or rabbit, then we can
select the appropriate database with `use_c_region_db()`:
```{r}
use_c_region_db("_IMGT.human.IGH+IGK+IGL.202412")
```

Calling `list_c_region_dbs()` again now should display an asterisk on
the very right end of the output to indicate the C-region db that
is currently selected.

See `?list_c_region_dbs` for more information.


# Use igblastn()


Now that we have selected a germline and C-region db to use with `igblastn()`,
we're almost ready to call `igblastn()`.

As mentioned previously, the `r Biocpkg("igblastr")` package includes a
FASTA file with 8437 pairs of human antibody sequences retrieved from OAS.
These are our _query sequences_ i.e. the immunoglobulin sequences that we're
going to analyse with `igblastn()`.

The path to our query sequences can be obtained with:
```{r}
query <- system.file(package="igblastr", "extdata", "1279067_1_Paired_sequences.fasta")
```

The `r Biocpkg("igblastr")` package also includes a JSON file that contains
metadata associated with our query sequences:
```{r}
json <- system.file(package="igblastr", "extdata", "1279067_1_Paired_All.json")
query_metadata <- jsonlite::fromJSON(json)
query_metadata
```
There are 8347 sequences, which are paired (i.e., both heavy and light chain
sequences are included). These sequences come from memory B cells from
peripheral blood mononuclear cell (PBMC) samples from one human individual
(age 38) with no particular disease or vaccination history.
The source for these sequences is the Jaffe et al. (2022) study, with DOI link
to the corresponding publication shown above.

Before we call `igblastn()`, we check the selected dbs by calling
`use_germline_db()` and `use_c_region_db()` with no argument:
```{r}
use_germline_db()
use_c_region_db()
```

Let's call `igblastn()`. Note that we're only interested in the best alignment
for each query sequence so we set `num_alignments_V` to 1. This will take up
to 3 min on an average laptop:
```{r}
out <- igblastn(query, num_alignments_V=1)
```
```{r, include=FALSE}
#out <- readRDS(file.path("precomputed_results", "out.rds"))
```

By default the output is a tibble in AIRR format:
```{r}
out
```

See `?igblastn` for more information.


# Downstream analysis examples


```{r, message = F}
library(ggplot2)
library(stringr)
library(dplyr)
library(ggseqlogo)
```

One common type of analysis of AIRR format data is to look at the distribution of percent mutation
in each sequence. For example, here we can look at percent mutation in the V
segments of each chain type (heavy, kappa, and lambda). Note that V percent mutation is 100 - v_identity.
```{r}
out %>%
  ggplot(aes(locus, 100 - v_identity)) +
  geom_point(position = position_jitter(width = 0.3), alpha = 0.1) +
  geom_boxplot(color = "blue", fill = NA, outliers = F, alpha = 0.3) +
  ggtitle("Distribution of V percent mutation by locus") +
  xlab(NULL)
```

Another common type of analysis is investigating the distribution of germline
genes (for example, V genes). Again, we want to stratify by locus or chain type.
```{r}
for (loc in c("IGH", "IGK", "IGL")) {
  
  df_v_gene <- out %>%
    filter(locus == loc) %>%
    mutate(v_gene = str_remove_all(v_call, "\\*[0-9]{2}")) %>% # drop allele info
    group_by(v_gene) %>%
    summarize(n = n(), .groups = "drop") %>%
    mutate(frac = n / sum(n))
  
  plt <- df_v_gene %>%
    ggplot(aes(frac, v_gene)) +
    geom_col() +
    scale_x_continuous('Percent of sequences', labels = scales::percent) +
    ylab("Germline gene") +
    ggtitle(paste0(loc, "V gene prevalence"))
  
  print(plt)
  
}
```

A third category of analyses investigates CDR3 sequences, including their
lengths and motifs (often visualized using sequence logo plots).
```{r}
out$cdr3_aa_length <- nchar(out$cdr3_aa)

out %>%
  group_by(locus, cdr3_aa_length) %>%
  summarize(n = n(), .groups = "drop") %>%
  ggplot(aes(cdr3_aa_length, n)) +
  facet_wrap(~locus) +
  geom_col() +
  ggtitle("Histograms of CDR3 length by locus")
```

```{r}
out %>%
  filter(locus == "IGK", cdr3_aa_length == 9) %>%
  pull(cdr3_aa) %>%
  ggseqlogo(method = "probability") +
  ggtitle("Logo plot of kappa chain CDR3 sequences that are 9 AA long")
```


# Future developments


TODO


# Session information


```{r}
sessionInfo()
```

